File: backend/app/schemas/ask.py
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator
from app.settings.schemas import LocalProviderOverride


KnowledgeScope = Literal["public", "private_sample", "mixed"]
AskStatus = Literal["ok", "provider_error", "invalid_input"]
MessageStatus = Literal["pending", "thinking", "streaming", "done", "error", "ok", "provider_error", "invalid_input"]
CitationSourceType = Literal["public_policy", "public_policy_demo", "private_sample", "private_upload"]


class AskCitation(BaseModel):
    doc_id: str
    knowledge_item_id: str | None = None
    title: str
    source_type: CitationSourceType = "public_policy"
    source: str
    source_url: str | None = None
    snippet: str
    chunk_id: str
    library_scope: Literal["personal", "shared"] | None = None
    file_id: str | None = None
    page_number: int | None = None
    sheet_name: str | None = None
    slide_number: int | None = None
    section_title: str | None = None


class AskSourceSummary(BaseModel):
    knowledge_scope: KnowledgeScope
    public_policy_count: int = 0
    public_policy_demo_count: int = 0
    private_sample_count: int = 0
    private_upload_count: int = 0
    total_citation_count: int = 0


class AskRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    question: str
    knowledge_scope: KnowledgeScope = "public"
    top_k: int = Field(default=5, ge=1)
    attached_file_ids: list[str] = Field(default_factory=list)
    attached_knowledge_item_ids: list[str] = Field(default_factory=list)
    request_group_id: str | None = None
    resume_cursor: int | None = Field(default=None, ge=0)
    provider_override: LocalProviderOverride | None = None

    @field_validator("question")
    @classmethod
    def normalize_question(cls, value: str) -> str:
        return value.strip()

    @field_validator("request_group_id")
    @classmethod
    def normalize_request_group_id(cls, value: str | None) -> str | None:
        if value is None:
            return None
        normalized = value.strip()
        return normalized or None


class AskResponse(BaseModel):
    answer: str
    mode: Literal["ask"]
    status: AskStatus
    citations: list[AskCitation] = Field(default_factory=list)
    source_summary: AskSourceSummary
    trace_id: str
    retrieval_trace: dict | None = None

File: frontend/src/types/ask.ts
export type AskStatus = "ok" | "provider_error" | "invalid_input";
export type KnowledgeScope = "public" | "private_sample" | "mixed";
export type CitationSourceType = "public_policy" | "public_policy_demo" | "private_sample" | "private_upload";
import type { LocalProviderOverride } from "./settings";

export type AskStreamLifecycleStatus = "pending" | "connecting" | "thinking" | "reconnecting" | "streaming" | "done" | "error" | "failed";
export type MessageStatus = AskStatus | AskStreamLifecycleStatus;
export type AskStreamEventName =
    | "message_start"
    | "status"
    | "thinking_delta"
    | "answer_delta"
    | "metadata"
    | "done"
    | "error";

export interface AskCitation {
    doc_id: string;
    knowledge_item_id?: string | null;
    title: string;
    source_type: CitationSourceType;
    source: string;
    source_url: string | null;
    snippet: string;
    chunk_id: string;
    library_scope?: "personal" | "shared" | null;
}

export interface AskSourceSummary {
    knowledge_scope: KnowledgeScope;
    public_policy_count: number;
    public_policy_demo_count?: number;
    private_sample_count: number;
    private_upload_count?: number;
    total_citation_count: number;
}

export interface AskRequest {
    question: string;
    knowledge_scope: KnowledgeScope;
    top_k: number;
    attached_file_ids?: string[];
    attached_knowledge_item_ids?: string[];
    request_group_id?: string;
    resume_cursor?: number;
    provider_override?: LocalProviderOverride;
}

export interface AskResponse {
    answer: string;
    mode: "ask";
    status: AskStatus;
    citations: AskCitation[];
    source_summary: AskSourceSummary;
    trace_id: string;
}

export interface AskStreamMessageStartEvent {
    user_message_id?: string | null;
    assistant_message_id?: string | null;
    trace_id?: string | null;
    request_group_id?: string | null;
    event_seq?: number | null;
}

export interface AskStreamStatusEvent {
    status: AskStreamLifecycleStatus;
    user_message_id?: string | null;
    assistant_message_id?: string | null;
    trace_id?: string | null;
    request_group_id?: string | null;
    attempt?: number | null;
    max_attempts?: number | null;
    recovered?: boolean | null;
    resume_supported?: boolean | null;
    provider_ref?: string | null;
    event_seq?: number | null;
}

export interface AskStreamDeltaEvent {
    delta?: string | null;
    text?: string | null;
    content?: string | null;
    synthetic?: boolean | null;
    user_message_id?: string | null;
    assistant_message_id?: string | null;
    trace_id?: string | null;
    request_group_id?: string | null;
    event_seq?: number | null;
}

export interface AskStreamMetadataEvent extends Partial<AskResponse> {
    user_message_id?: string | null;
    assistant_message_id?: string | null;
    thinking_content?: string | null;
    memory_state?: {
        context_usage_estimate: number;
        context_budget_estimate: number;
        summary_present: boolean;
        summary_updated_at?: string | null;
        compacted_message_count: number;
        compaction_status: "idle" | "compacted" | "failed";
        summary_estimated_tokens?: number;
    } | null;
    context_source?: {
        recent_message_count: number;
        summary_present: boolean;
        citation_count: number;
    } | null;
    request_group_id?: string | null;
    title_updated?: boolean | null;
    provider_ref?: string | null;
    event_seq?: number | null;
}

export interface AskStreamErrorEvent {
    message?: string | null;
    detail?: string | null;
    status?: AskStatus | null;
    trace_id?: string | null;
    user_message_id?: string | null;
    assistant_message_id?: string | null;
    request_group_id?: string | null;
    attempt?: number | null;
    max_attempts?: number | null;
    event_seq?: number | null;
}

export interface AskStreamCallbacks {
    onMessageStart?: (event: AskStreamMessageStartEvent) => void;
    onStatus?: (event: AskStreamStatusEvent) => void;
    onThinkingDelta?: (event: AskStreamDeltaEvent) => void;
    onAnswerDelta?: (event: AskStreamDeltaEvent) => void;
    onMetadata?: (event: AskStreamMetadataEvent) => void;
    onDone?: (event: AskStreamMetadataEvent) => void;
    onError?: (event: AskStreamErrorEvent) => void;
}
